### PR TITLE
HMS-5125: Add support for resumable downloads

### PR DESCRIPTION
**Pull request by Andrewgdewar**
## Summary
Adds support for resumable downloads. 

Can test with [this](https://github.com/content-services/content-sources-frontend/pull/397) front-end PR.

## Testing steps

- Create an upoad repository.
- Select an item for upload, allow it to upload around halfway and then immediately hit refresh.
- After refresh, choose the same item, you should be able to see that item continue where you left off. 
- After this item has been completed, upload it, and create another upload repository. 
- Choose the same file, it should instantly be available.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-5125

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/Fixes 5125

**Review by dominikvagner — CHANGES_REQUESTED** [review]
small change needed in the API spec, otherwise looks good, nice job! 😄
everything works as expected 👍🏼 

_(tiny nit about style/formatting: in Go the `if err != nil` check usually follow the err assignment without a new line in between them, makes it more clear what error they are checking)_

File: pkg/api/pulp.go [hunk 1]
```diff
@@ -1,7 +1,9 @@
 package api
 
 type CreateUploadRequest struct {
-	Size int64 `json:"size" validate:"required"` // Size of the upload in bytes
+	Size      int64  `json:"size" validate:"required"`      // Size of the upload in bytes
+	ChunkSize int64  `json:"chunkSize" validate:"required"` // Size of the chunk
```
[the remark this answers is not in the document]
Comment (Andrewgdewar) [reply pkg/api/pulp.go:5]: Thanks buddy!

**Review by dominikvagner — APPROVED** [review]

File: pkg/dao/uploads.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	return nil
+}
+
+func (t uploadDaoImpl) GetExistingUploadIDAndCompletedChunks(ctx context.Context, orgID string, sha256 string, chunkSize int64) (string, []string, error) {
+	db := t.db.Model(models.Upload{}).WithContext(ctx)
+
+	var result models.Upload
+
+	db.Where("org_id = ?", orgID).Where("chunk_size = ?", chunkSize).Where("sha256 = ?", sha256).First(&result)
```
Comment (jlsherrill) [inline pkg/dao/uploads.go:46]: i'd probably add an index on org_id, chunk_size, and sha256 to the db migration  to make this a bit more efficient

Comment (Andrewgdewar) [reply pkg/dao/uploads.go:46]: Added, please take a look for a sanity check 😓 !

**Pull request merged**